File: tests/stateMachine.test.ts
import {
    verifyInput,
    verifyStates,
    makeStateMachine,
    Alphabet,
    States
} from "../library/StateMachineTS";

describe("TS verifyInput", () => {
    const alphabet: Alphabet = ["0", "1"];

    it("should return the same string if valid", () => {
        expect(verifyInput("1010", alphabet)).toBe("1010");
    });

    it("should throw if input contains invalid characters", () => {
        expect(() => verifyInput("1020", alphabet)).toThrow('Bit "2" at position 2 is not in alphabet');
    });

    it("should throw if input is empty", () => {
        expect(() => verifyInput("", alphabet)).toThrow("Input must not be empty");
    });

    it("should convert numeric input to string", () => {
        expect(verifyInput(101, alphabet)).toBe("101");
    });
});

describe("verifyStates", () => {
    const alphabet: Alphabet = ["0", "1"];

    it("should validate proper state definitions", () => {
        const states: States<string> = {
            S0: { "0": "S0", "1": "S1", result: "A" },
            S1: { "0": "S2", "1": "S0", result: "B" },
            S2: { "0": "S1", "1": "S2", result: "C" }
        };
        expect(() => verifyStates(states, alphabet)).not.toThrow();
    });

    it("should throw if a transition points to an undefined state", () => {
        const badStates: States<string> = {
            S0: { "0": "S1", result: "A" }
        };
        expect(() => verifyStates(badStates, alphabet)).toThrow('State "S1" in state "S0" is not defined');
    });

    it("should throw if result key is missing", () => {
        const badStates: any = {
            S0: { "0": "S0", "1": "S0" }
        };
        expect(() => verifyStates(badStates, alphabet)).toThrow('State "S0" must have a "result"');
    });
});

describe("makeStateMachine", () => {
    const alphabet: Alphabet = ["0", "1"];
    const states: States<string> = {
        S0: { "0": "S0", "1": "S1", result: "A" },
        S1: { "0": "S2", "1": "S0", result: "B" },
        S2: { "0": "S1", "1": "S2", result: "C" }
    };
    const fsm = makeStateMachine(states, alphabet);

    it("should compute the correct result from a binary string", () => {
        expect(fsm("011", "S0")).toBe("A");
        expect(fsm("01", "S0")).toBe("B");
        expect(fsm("010", "S0")).toBe("C");
    });

    it("should throw for invalid input character", () => {
        expect(() => fsm("012", "S0")).toThrow('Bit "2" at position 2 is not in alphabet');
    });
});

File: tests/state_machine_test.go
package tests

import (
	"testing"

	"modulo3/library"
)

func TestVerifyInput(t *testing.T) {
	alphabet := library.Alphabet{"0", "1"}

	_, err := library.VerifyInput("102", alphabet)
	if err == nil {
		t.Error("Expected error for invalid input")
	}

	_, err = library.VerifyInput("", alphabet)
	if err == nil {
		t.Error("Expected error for empty input")
	}

	val, err := library.VerifyInput("1010", alphabet)
	if err != nil || val != "1010" {
		t.Error("Valid input failed")
	}
}

func TestVerifyStates(t *testing.T) {
	alphabet := library.Alphabet{"0", "1"}

	goodStates := library.States[string]{
		"S0": {Transitions: map[string]string{"0": "S0", "1": "S1"}, Result: library.Ptr("A")},
		"S1": {Transitions: map[string]string{"0": "S2", "1": "S0"}, Result: library.Ptr("B")},
		"S2": {Transitions: map[string]string{"0": "S1", "1": "S2"}, Result: library.Ptr("C")},
	}

	if err := library.VerifyStates(goodStates, alphabet); err != nil {
		t.Errorf("Expected valid states, got error: %v", err)
	}

	badStates := library.States[string]{
		"S0": {Transitions: map[string]string{"0": "S1"}, Result: nil},
	}

	if err := library.VerifyStates(badStates, alphabet); err == nil {
		t.Error("Expected error for missing result or undefined state")
	}
}

func TestMakeStateMachine(t *testing.T) {
	alphabet := library.Alphabet{"0", "1"}

	states := library.States[string]{
		"S0": {Transitions: map[string]string{"0": "S0", "1": "S1"}, Result: library.Ptr("A")},
		"S1": {Transitions: map[string]string{"0": "S2", "1": "S0"}, Result: library.Ptr("B")},
		"S2": {Transitions: map[string]string{"0": "S1", "1": "S2"}, Result: library.Ptr("C")},
	}

	fsm := library.MakeStateMachine(states, alphabet)

	result, err := fsm("011", "S0")
	if err != nil || result != "A" {
		t.Errorf("Expected result 'A', got %v (err: %v)", result, err)
	}

	_, err = fsm("012", "S0")
	if err == nil {
		t.Error("Expected error for invalid input character")
	}
}
